Userland/Libraries/LibAudio/FlacLoader.cpp

```diff
@@ -0,0 +1,899 @@
+        if (m_file) {
+            fis = Core::InputFileStream(*m_file);
+            return InputBitStream(*fis);
+        }
+        return InputBitStream(m_stream->get<InputMemoryStream>());
+    }();
+
+#define CHECK_OK(msg)                                                      \
+    do {                                                                   \
+        if (!ok) {                                                         \
+            m_error_string = String::formatted("Parsing failed: {}", msg); \
+            return {};                                                     \
+        }                                                                  \
+    } while (0)
+
+    // Magic number
+    u32 flac = bit_input.read_bits_big_endian(32);
+    m_data_start_location += 4;
+    ok = ok && flac == 0x664C6143; // "flaC"
+    CHECK_OK("FLAC magic number");
+
+    // Receive the streaminfo block
+    FlacRawMetadataBlock streaminfo = next_meta_block(bit_input);
+    // next_meta_block sets the error string if something goes wrong
+    ok = ok && m_error_string.is_empty();
+    CHECK_OK(m_error_string);
+    ok = ok && (streaminfo.type == FlacMetadataBlockType::STREAMINFO);
+    CHECK_OK("First block type");
+    InputMemoryStream streaminfo_data_memory(streaminfo.data.bytes());
+    InputBitStream streaminfo_data(streaminfo_data_memory);
+
+    // STREAMINFO block
+    m_min_block_size = streaminfo_data.read_bits_big_endian(16);
+    ok = ok && (m_min_block_size >= 16);
+    CHECK_OK("Minimum block size");
+    m_max_block_size = streaminfo_data.read_bits_big_endian(16);
+    ok = ok && (m_max_block_size >= 16);
+    CHECK_OK("Maximum block size");
+    m_min_frame_size = streaminfo_data.read_bits_big_endian(24);
+    m_max_frame_size = streaminfo_data.read_bits_big_endian(24);
+    m_sample_rate = streaminfo_data.read_bits_big_endian(20);
+    ok = ok && (m_sample_rate <= 655350);
+    CHECK_OK("Sample rate");
+    m_num_channels = streaminfo_data.read_bits_big_endian(3) + 1; // 0 ^= one channel
+
+    u8 bits_per_sample = streaminfo_data.read_bits_big_endian(5) + 1;
+    if (bits_per_sample == 8) {
+        // FIXME: Signed/Unsigned issues?
+        m_sample_format = PcmSampleFormat::Uint8;
+    } else if (bits_per_sample == 16) {
+        m_sample_format = PcmSampleFormat::Int16;
+    } else if (bits_per_sample == 24) {
+        m_sample_format = PcmSampleFormat::Int24;
+    } else if (bits_per_sample == 32) {
+        m_sample_format = PcmSampleFormat::Int32;
+    } else {
+        ok = false;
+        CHECK_OK("Sample bit depth");
+    }
+
+    m_total_samples = streaminfo_data.read_bits_big_endian(36);
+    // Parse checksum into a buffer first
+    ByteBuffer md5_checksum = ByteBuffer::create_uninitialized(128 / 8);
+    streaminfo_data.read(md5_checksum);
+    // HACK: There must be a better way. Also Endianness
+    for (unsigned int i = 0; i < md5_checksum.size(); ++i) {
+        m_md5_checksum[i] = md5_checksum[i];
+    }
```

Comment:
Why not just `m_md5_checksum = move(md5_checksum);`?

Reply:
Because I can't C++ :)

Reply:
```cpp
../../Userland/Libraries/LibAudio/FlacLoader.cpp:130:20: error: incompatible types in assignment of ‘AK::Detail::ByteBuffer<32>’ to ‘u8 [16]’ {aka ‘unsigned char [16]’}
  130 |     m_md5_checksum = move(md5_checksum);
      |     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
```

Reply:
Types are incompatible, while md5_checksum is a byte buffer for reading from the stream, m_md5_checksum is an array because the size is constant.

Reply:
is there any reasoning against `m_md5_checksum` being an `ByteBuffer` instead of a raw array? it could help make the code a lot more idiomatic.

Reply:
more performance?

Reply:
Ah, given that it's a normal array, the fastest thing you can do is to memcpy it in.

Reply:
In slightly less annoying code:
```c++
md5_checksum.bytes().copy_to({ m_md5_checksum, sizeof(m_md5_checksum) });
```